### PR TITLE
Websocket server implementation

### PR DESCRIPTION
**Pull request by vuvkar**

### CODE REVIEW

File: editor/src/com/talosvfx/talos/editor/addons/scene/assets/AssetRepository.java [hunk 1]
```diff
@@ -171,8 +171,12 @@ public void run () {
 	private void loadChangesFromScripts(Function<Path, FileVisitResult> function) {
 
 		FileHandle exportedScriptsFolderHandle = getExportedScriptsFolderHandle();
+		if (!exportedScriptsFolderHandle.exists()) {
+			return;
```
Comment (vuvkar) [inline editor/src/com/talosvfx/talos/editor/addons/scene/assets/AssetRepository.java:175]: This check is for projects which isn't in the directory of tinyGdx runtime

File: editor/src/com/talosvfx/talos/editor/addons/scene/events/ComponentUpdated.java [hunk 2]
```diff
@@ -1,27 +1,40 @@
...
 import com.talosvfx.talos.editor.notifications.Notifications;
 
+import java.awt.geom.NoninvertibleTransformException;
+
 public class ComponentUpdated implements Notifications.Event {
 
     private AComponent component;
     private boolean rapid;
 
+    private boolean notifyUI;
```
Comment (vuvkar) [inline editor/src/com/talosvfx/talos/editor/addons/scene/events/ComponentUpdated.java:16]: I hate to do this :/

File: editor/src/com/talosvfx/talos/editor/addons/scene/utils/scriptProperties/ScriptMetadataParser.java [hunk 3]
```diff
@@ -122,8 +122,8 @@ private void processLine(String line, ScriptMetadata metadata) throws ScriptMeta
             if (keyWordsCount % 2 != 0) {
                 throw new ScriptMetadataParserException("Invalid contruction of type or arguments");
             }
-            if (keyWordsCount < 6) {
-                throw new ScriptMetadataParserException("The field musy have  [name], [type] and [defaultValue] parameters");
+            if (keyWordsCount < 4) {
+                throw new ScriptMetadataParserException("The field must have  [name] and [type] parameters");
```
Comment (vuvkar) [inline editor/src/com/talosvfx/talos/editor/addons/scene/utils/scriptProperties/ScriptMetadataParser.java:126]: so the default value isn't must actually

File: editor/src/com/talosvfx/talos/editor/notifications/Notifications.java [hunk 4]
```diff
@@ -36,6 +33,22 @@ public interface EventRunner {
...
         @Override
         void reset ();
+
+        default boolean notifyThroughSocket () {
+            return false;
+        }
+
+        default String getEventType () {
+            return "";
+        }
```
Comment (vuvkar) [inline editor/src/com/talosvfx/talos/editor/notifications/Notifications.java:43]: Maybe create an Event abstract class, make all events inherit from this one, and have the methods there?
I just don't like having default methods in interfaces

File: editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java [hunk 5]
```diff
@@ -29,6 +30,8 @@
 	private Supplier<T> supplier;
 	private ValueChanged<T> valueChanged;
 
+	private Object parent;
```
Comment (vuvkar) [inline editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java:33]: the connection between property widget and AComponent is lost once the widget is created. Any ideas for better system?

File: editor/src/com/talosvfx/talos/editor/addons/scene/SceneEditorWorkspace.java [hunk 6]
```diff
@@ -43,6 +44,7 @@
 import com.talosvfx.talos.editor.widgets.ui.ViewportWidget;
 import com.talosvfx.talos.runtime.ParticleEffectDescriptor;
 
+import java.awt.*;
```
Comment (Tom-Ski) [inline editor/src/com/talosvfx/talos/editor/addons/scene/SceneEditorWorkspace.java:47]: Bad import?

File: editor/src/com/talosvfx/talos/editor/addons/scene/events/ComponentUpdated.java [hunk 7]
```diff
@@ -1,27 +1,40 @@
 package com.talosvfx.talos.editor.addons.scene.events;
 
+import com.badlogic.gdx.utils.Json;
+import com.talosvfx.talos.editor.addons.scene.SceneEditorWorkspace;
+import com.talosvfx.talos.editor.addons.scene.logic.GameObject;
 import com.talosvfx.talos.editor.addons.scene.logic.components.AComponent;
 import com.talosvfx.talos.editor.notifications.Notifications;
 
+import java.awt.geom.NoninvertibleTransformException;
```
Comment (Tom-Ski) [inline editor/src/com/talosvfx/talos/editor/addons/scene/events/ComponentUpdated.java:9]: eh?

File: editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java (hunk 5 above)
Comment (Tom-Ski) [reply editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java:33]: Dont get this? Why is it lost?

**Review by Tom-Ski — CHANGES_REQUESTED** [review]

File: editor/src/com/talosvfx/talos/editor/addons/scene/SceneEditorWorkspace.java (hunk 6 above)
Comment (vuvkar) [reply editor/src/com/talosvfx/talos/editor/addons/scene/SceneEditorWorkspace.java:47]: yep, reverted with next commit

File: editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java (hunk 5 above)
Comment (vuvkar) [reply editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java:33]: property widget is connected to the actual fields with reflection. So it doesn't know whose fields its changing. I had to keep the parent here and pass it to the change event, in order to send it to socket

Comment (Tom-Ski) [reply editor/src/com/talosvfx/talos/editor/widgets/propertyWidgets/PropertyWidget.java:33]: ah, thats fine

**Review by Tom-Ski — APPROVED** [review]

**Pull request merged**